### DISCUSSION TITLE
how to secondary window?

### DISCUSSION THREAD

**Discussion by hemna** [beeware/toga#2929] [General]
I'm writing a chat app that is going to mimick the android and ios sms app.  The app is for ham radio using APRS network.

I'm just starting to write the app and riding the steep learning curve for Toga.   The chat app has a main list of discussions with other ham callsigns, a new button to start a new chat with someone.  The problem is I don't know how to manage the visible window on ios/android.   

clicking/tapping on the new button from the main chat list, opens a new window that contains
To: <callsign>
Message: <the message text>
<Send Button> <Cancel Button>

The problem I just ran into on iOS is this error.

```
    raise RuntimeError("Secondary windows cannot be created on iOS")
RuntimeError: Secondary windows cannot be created on iOS 
```

It would be nice if there was some real examples that shows how to manage windows for the mobile platforms.   This seems like a lot of trial and error on my part on how to write/structure this app correctly.

I have seen the examples for widgets, but nothing to show how to write a real world app that has multiple other windows (settings, login, about) as well as how to use icons as buttons.

Thank in advance and I apoligize for my newbness.

**Comment by hemna** [comment]
![Screenshot 2024-10-23 at 12 56 25 PM](https://github.com/user-attachments/assets/120befa9-73c5-47e8-81b1-24fcd9e891ff)

**Comment by freakboy3742** [comment]
The short answer is that at present, you *don't* use secondary windows on mobile. That's just a limitation of the platform.

The best workaround at present is to manipulate the root content of the window - change `main_window.content` between your "actual" main window content, and your "dialog" or secondary window content. I agree this is messy; but it's the best I can offer right now.

We're very aware that "high level app navigation" is an issue that Toga doesn't currently deal with well; we're hoping to be looking at this problem soon. We're also aware that there's a lack of tutorials and documentation for some of these high level tasks; that's the the unfortunate price for a toolkit that is still at v0.4.

**Comment by hemna** [comment]
So, I'm creating a new toga.Box object with all the content, and then calling 
self.main_window.content = my_new_box

and nothing happens


```
    def switch_windows(self, new_box: toga.Box):
        """Change the main window content to the new window."""
        if not self.prev_content:
            self.prev_content = self.main_window.content
            print(f"Setting prev_content to {self.prev_content}")
            print(f"Setting new content to {new_box}")
            self.main_window.content = new_box
            self.main_window.show()

    def reset_windows(self):
        """Reset the main window content to the previous content."""
        if self.prev_content:
            self.main_window.content = self.prev_content
            self.prev_content = None
            self.main_window.show()
           
```

```
Setting prev_content to <Box:0x10debee50>.png; falling back to default icon
Setting new content to <NewChatWindow:0x10e35f670>
```

No change in the window


Here is my NewChatWindow class for the most part.  This used to be a toga.Window child, but that didn't work either.

```
class NewChatWindow(toga.Box):
    def __init__(self, callback):
        self.callback = callback
        super().__init__()

        self.to_input = toga.TextInput(style=Pack(flex=1, width=100, padding=PADDING))
        self.msg_input = toga.TextInput(style=Pack(flex=1, width=300, padding=PADDING))
        self.ok_button = toga.Button(
            "Send",
            on_press=self.on_accept,
            style=Pack(padding=PADDING)
        )
        self.cancel_button = toga.Button(
            "Cancel",
            on_press=self.on_cancel,
            style=Pack(padding=PADDING)
        )

        message_label = toga.Label(
            "New callsign Chat",
            style=Pack(alignment=CENTER, padding=PADDING),
        )
        msg_to_box = toga.Box(
            children=[
                toga.Label("To Callsign:"),
                self.to_input,
            ],
            style=Pack(flex=1, direction=ROW, padding=PADDING, alignment=LEFT),
        )
        msg_input_box = toga.Box(
            children=[
                toga.Label("Message   :"),
                self.msg_input,
            ],
            style=Pack(flex=1, direction=ROW, padding=PADDING, alignment=LEFT),
        )

        button_box = toga.Box(
            children=[
                self.ok_button,
                self.cancel_button
            ],
            style=Pack(flex=1, direction=ROW, alignment=CENTER, padding=PADDING),
        )

        self.content = toga.Box(
            children=[
                message_label,
                msg_to_box,
                msg_input_box,
                button_box,
            ],
            style=Pack(direction=COLUMN, padding=PADDING),
        )

    def on_accept(self, widget, **kwargs):
        new_msg = {
            "to": self.to_input.value,
            "msg": self.msg_input.value,
        }
        self.callback(new_msg)
        self.close()

    def on_cancel(self, widget, **kwargs):
        self.close()
```

**Reply by mhsmith** [reply]
`Box` doesn't have a `content` attribute, so assigning to `self.content` will have no effect. Instead, you should pass a `children` list when you call `super().__init__()`.

**Comment by hemna** [comment]
I'm just not sure how to accomplish this.  I need a new window that acts like a dialog that gets input and has a send(save/Ok) button and a cancel button.   Get the values of those inputs and then change the main window contents back to what it was previously.

I'll need this for a settings page, about page, chat contents page (where the actual chat messages are displayed between the user and remote user), etc...

**Comment by hemna** [comment]
I also tried setting the main_window.content  to 

```
new_box = toga.Box(
                children=[
                    toga.Label("New Chat", style=Pack(flex=1, padding=PADDING)),
                ],
                style=Pack(direction=COLUMN, padding=PADDING),
            )
```

also nothing happens.

**Reply by freakboy3742** [reply]
As you've only provided selected snippets of code, rather than a complete example, it's difficult to say for certain where your code is going wrong. My guess is that you're treating the Window as "main_window.content", which won't work - you need to assign the *content* of the window. Creating a subclass of Window *at all* is going to be problematic.

I can confirm that the following works:

```
class Test2929(toga.App):
    def startup(self):
        self.main_box = toga.Box(children=[toga.Button("Switch", on_press=self.switch)])
        self.other_box = toga.Box(
            children=[toga.Button("Revert", on_press=self.revert)]
        )
        self.main_window = toga.MainWindow(title=self.formal_name)
        self.main_window.content = self.main_box
        self.main_window.show()

    def switch(self, widget, **kwargs):
        self.main_window.content = self.other_box

    def revert(self, widget, **kwargs):
        self.main_window.content = self.main_box
```
When the app starts, it will show a "Switch" button; press that button, and it will be replaced with a "Revert" button. Press that button, and you'll get back to the "Switch" button.

**Comment by l-o-u-i-s-g** [comment]
I made a sample example a while ago on how to navigate between boxes,  I just uploaded the code for you: 
https://github.com/l-o-u-i-s-g/toga-snippets/tree/main/box-navigation
